File: C/65_valid_number.c
#include <stdbool.h>
#include <stdio.h>

bool isNumber(char* s) {
    bool exp, dot;
    int digidx, x, fsm_state;
    bool signs[] = {false, false};
    int digits[] = {0, 0};

    if (!s || !(*s))
        return false;
    exp = false;
    dot = false;
    fsm_state = 0; /* 0 - sign state (initial), 1 - number state */
    digidx = 0;
    while (*s) {
        x = *s - '0';
        if (fsm_state) {
            if (!exp && (*s == 'e' || *s == 'E')) {
                exp = true;
                digidx = 1;
                dot = true; /* exponent is always an integer */
                fsm_state = 0; /* return back to the 'sign' state */
            } else if (!dot && *s == '.') {
                dot = true;
            } else if (x >= 0 && x < 10) {
                digits[digidx] += 1;
            } else
                return false;
        } else {
            if (*s == '-' || *s == '+') {
                signs[digidx] = true;
                fsm_state = 1;
            } else if (x >= 0 && x < 10) {
                digits[digidx] += 1;
                fsm_state = 1;
            } else if (!dot && *s == '.') {
                dot = true;
                fsm_state = 1;
            } else
                return false;
        }
        s++;
    }
    if ((exp && !digits[1]) || (dot && !digits[0]) || (signs[0] && !digits[0]))
        return false;
    return true;
}

int main(int argc, char **argv) {
     bool res;
     char *vect[] = {NULL, "", " ", ".", "+", "-", "e", "e1", "e-", "+e2", "1.0e+5", "1.0", "1", "-232", "1212.2e4", "4.66E-5"};

     for (int i = 0; i < sizeof(vect) / sizeof(vect[0]); i++) {
         res = isNumber(vect[i]);
	 printf("'%s' %d\n", vect[i], res);
     }
     return 0;
}

File: 65_valid_number.py
class Solution(object):
    def isNumber(self, s):
        """
        :type s: str
        :rtype: bool
        """
        if s is None or len(s) == 0:
            return False

        fsm_state = 0 # 0 - sign { {+, -} (0..1) || digit (0..*) || . (0..1) },
                      # 1 - number { digit (0..*) || . (0.. 1) || {e, E} (0..1) }
        exp = False
        dot = False
        signs = [False, False]
        digits = [0, 0]
        digidx = 0 
        for c in s:
            x = ord(c) - ord('0')
            if fsm_state == 1:
                if not exp and c in 'eE':
                    exp = True
                    digidx = 1
                    dot = True # exponent is always an integer
                    fsm_state = 0
                elif not dot and c == '.':
                    dot = True
                elif 0 <= x <= 9:
                    digits[digidx] += 1
                else:
                    return False
            elif fsm_state == 0:
                if c in '-+':
                    signs[digidx] = True
                    fsm_state = 1
                elif 0 <= x <= 9:
                    digits[digidx] += 1
                    fsm_state = 1
                elif not dot and c == '.':
                    dot = True
                    fsm_state = 1
                else:
                    return False
        if (exp and digits[1] == 0) or (dot and digits[0] == 0) \
                or (signs[0] and digits[0] == 0):
            return False
        return True

vectors = [
        None, False,
        "", False,
        ".", False,
        "e", False,
        "+", False,
        "-", False,
        "-e", False,
        ".e1", False,
        "2", True,
        "0089", True,
        "-0.1", True,
        "+3.14", True,
        "4.", True,
        "-9.", True,
        "2e10", True,
        "-90E3", True,
        "3e+7", True,
        "+6e-1", True,
        "53.5e93", True,
        "-123.456e789", True,
        "abc", False,
        "1a", False,
        "1e", False,
        "e3", False,
        "99e2.5", False,
        "--6", False,
        "-+3", False,
        "95a54e53", False
        ]

for i in range(0, len(vectors), 2):
    s = vectors[i]
    expected = vectors[i + 1]
    print(f'{s}')
    returned = Solution().isNumber(s)
    assert expected == returned, f'for \'{s}\' expected {expected}, but returned {returned}!'
